app/routes.py

```diff
@@ -1,2 +1,344 @@
+                }
+                }
+                ), 200
+
+
+@tasks_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"])
+def patch_it_down(task_id):
+    task = Task.query.get(task_id)
+
+    if task is None:
+        return jsonify(None), 404
+
+    if request.method == "PATCH":
+        task.completed_at = None
+        task.is_complete = False
+        db.session.commit()
+
+
+        return jsonify(
+            {
+            "task":
+            {
+            "id" : task.task_id,
+            "title" : task.title,
+            "description" : task.description,
+            "is_complete" : task.is_complete
+            }
+            }
+            ), 200
+
+
+
+@goals_bp.route("", methods=["GET", "POST"])
+def handle_goals():
+
+    request_body = request.get_json()
+    goals = Goal.query.all()
+
+    if request.method == "GET":
+ 
+        goals_response = []
+        for goal in goals:
+            goals_response.append(
+                {
+                    "id" : goal.goal_id,
+                    "title" : goal.title
+                }
+            )
+
+        return jsonify(goals_response), 200
+
+    elif request.method == "POST":
+        request_body = request.get_json()
+
+        if "title" not in request_body:
+            return jsonify(
+                {
+                "details": "Invalid data"
+                }
+                ), 400
+
+        new_goal = Goal(title=request_body["title"])
+
+        db.session.add(new_goal)
+        db.session.commit()
+
+        return {
+                "goal": { "id": new_goal.goal_id,
+                "title": new_goal.title
+                }
+                }, 201
+        
+
+@goals_bp.route("/<goal_id>", methods=["GET", "PUT", "DELETE"])
+def handle_goal(goal_id):
+    goal = Goal.query.get(goal_id)
+
+    if goal is None:
+        return jsonify(None), 404
+
+    if request.method == "GET":
+        return {"goal":
+            {
+            "id" : goal.goal_id,
+            "title" : goal.title
+                }
+            }, 200
+
+    elif request.method == "PUT":
+        request_body = request.get_json()
+
+        goal.title = request_body["title"]
+
+        db.session.commit()
+
+        return {
+                "goal": {"id": goal.goal_id, "title": goal.title
+                }
+                }, 200
+    elif request.method == "DELETE":
+        db.session.delete(goal)
+        db.session.commit()
+        return jsonify(
+            {
+                'details': (f'Goal {goal.goal_id} \"{goal.title}\" successfully deleted')
+            }
+
+        ), 200
+
+@goals_bp.route("/<goal_id>/tasks", methods=["POST", "GET"])
+def goals_and_tasks(goal_id):
+    goal = Goal.query.get(goal_id)
+    request_body = request.get_json()
+
+    if not goal:
+        return jsonify(None), 404
+
+    if request.method == "POST":
+        
+        the_list = []
+        for num in request_body["task_ids"]:
+            the_task = Task.query.get(num)
```

Comment:
What would happen if any of the tasks we looked up here were invalid ids (or not numeric ids at all)? We should make sure that the task ids actually load a valid task, otherwise the following line will try to access a `NoneType` object.